mongofiles/mongofiles.go

```diff
@@ -130,228 +154,287 @@ func (mf *MongoFiles) ValidateCommand(args []string) error {
 }
 
 // Query GridFS for files and display the results.
-func (mf *MongoFiles) findAndDisplay(gfs *mgo.GridFS, query bson.M) (string, error) {
-	display := ""
-
-	cursor := gfs.Find(query).Iter()
-	defer cursor.Close()
-
-	var file GFSFile
-	for cursor.Next(&file) {
-		display += fmt.Sprintf("%s\t%d\n", file.Name, file.Length)
-	}
-	if err := cursor.Err(); err != nil {
+func (mf *MongoFiles) findAndDisplay(query bson.M) (string, error) {
+	gridFiles, err := mf.getGFSFiles(query)
+	if err != nil {
 		return "", fmt.Errorf("error retrieving list of GridFS files: %v", err)
 	}
 
+	var display string
+	for _, gridFile := range gridFiles {
+		display += fmt.Sprintf("%s\t%d\n", gridFile.Name, gridFile.Length)
+	}
+
 	return display, nil
 }
 
 // Return the local filename, as specified by the --local flag. Defaults to
 // the GridFile's name if not present. If GridFile is nil, uses the filename
 // given on the command line.
-func (mf *MongoFiles) getLocalFileName(gridFile *mgo.GridFile) string {
+func (mf *MongoFiles) getLocalFileName(gridFile *gfsFile) string {
 	localFileName := mf.StorageOptions.LocalFileName
 	if localFileName == "" {
 		if gridFile != nil {
-			localFileName = gridFile.Name()
+			localFileName = gridFile.Name
 		} else {
 			localFileName = mf.FileName
 		}
 	}
 	return localFileName
 }
 
-// handle logic for 'get' command
-func (mf *MongoFiles) handleGet(gfs *mgo.GridFS) error {
-	gFile, err := gfs.Open(mf.FileName)
+// handleGet contains the logic for the 'get' and 'get_id' commands
+func (mf *MongoFiles) handleGet() (err error) {
+	file, err := mf.getTargetGFSFile()
 	if err != nil {
-		return fmt.Errorf("error opening GridFS file '%s': %v", mf.FileName, err)
+		return err
 	}
-	defer gFile.Close()
-	if err = mf.writeFile(gFile); err != nil {
+	defer func() {
+		closeErr := file.Close()
+		if closeErr != nil {
```

Comment:
This is actually an outdated patch-I've since changed this (and all other `defer` + `Close`s) to use the `DeferredCloser` example you sent in your previous review. Is that still valid, or would you prefer the simpler approach you outlined here?